sdks/java/core/src/main/java/org/apache/beam/sdk/util/UnboundedScheduledExecutorService.java

```diff
@@ -0,0 +1,504 @@
+            // it in another ScheduledFutureTask.
+            threadPoolExecutor.execute(command);
+          }
+        };
+    this.tasks = new PriorityQueue<>();
+    this.launchTasks =
+        threadPoolExecutor.submit(new TaskLauncher(tasks, threadPoolExecutor, clock));
+  }
+
+  private static class TaskLauncher implements Callable<Void> {
+    private final PriorityQueue<ScheduledFutureTask<?>> tasks;
+    private final ThreadPoolExecutor threadPoolExecutor;
+    private final NanoClock clock;
+
+    private TaskLauncher(
+        PriorityQueue<ScheduledFutureTask<?>> tasks,
+        ThreadPoolExecutor threadPoolExecutor,
+        NanoClock clock) {
+      this.tasks = tasks;
+      this.threadPoolExecutor = threadPoolExecutor;
+      this.clock = clock;
+    }
+
+    @Override
+    public Void call() throws Exception {
+      while (true) {
+        synchronized (tasks) {
+          if (threadPoolExecutor.isShutdown()) {
+            return null;
+          }
+          ScheduledFutureTask<?> task = tasks.peek();
+          if (task == null) {
+            tasks.wait();
+            continue;
+          }
+          long nanosToWait = LongMath.saturatedSubtract(task.time, clock.nanoTime());
+          if (nanosToWait > 0) {
+            long millisToWait = nanosToWait / 1_000_000;
+            int nanosRemainder = (int) (nanosToWait % 1_000_000);
+            tasks.wait(millisToWait, nanosRemainder);
+            continue;
+          }
+          // Remove the task from the queue since it is ready to be scheduled now
+          task = tasks.remove();
+          threadPoolExecutor.execute(task);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    threadPoolExecutor.shutdown();
+    synchronized (tasks) {
+      // Notify tasks which checks to see if the ThreadPoolExecutor is shutdown and exits cleanly.
+      tasks.notify();
+    }
+
+    // Re-throw any errors during shutdown of the launchTasks thread.
+    try {
+      launchTasks.get();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e.getCause());
+    }
+  }
+
+  @Override
+  public List<Runnable> shutdownNow() {
+    shutdown();
+    synchronized (tasks) {
+      List<Runnable> rval = new ArrayList<>(tasks);
+      tasks.clear();
+      rval.addAll(threadPoolExecutor.shutdownNow());
+      return rval;
+    }
+  }
+
+  @Override
+  public boolean isShutdown() {
+    return threadPoolExecutor.isShutdown();
+  }
+
+  @Override
+  public boolean isTerminated() {
+    return threadPoolExecutor.isTerminated();
+  }
+
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    return threadPoolExecutor.awaitTermination(timeout, unit);
+  }
+
+  @Override
+  public void execute(Runnable command) {
```

Comment:
I assume we can't use an @nonnull annotation because we are implementing a method?

Reply:
`@nonnull` only adds checking for those that use that tooling when compiling their pipeline.

The explicit check makes sure `command` isn't `null` for all those that don't use findbugs/spotbugs and to match the ScheduledExecutorService method contracts.

Reply:
makes sense